In [4]:
# =============================================================================
# STEP 1: LOAD PROCESSED DATASET
# =============================================================================

import pandas as pd
import numpy as np


# Load cleaned dataset
df = pd.read_csv(
    "../../../data/processed/diabetes_cleaned.csv"
)


print("=" * 70)
print("PROCESSED DATASET")
print("=" * 70)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")


display(df.head())

PROCESSED DATASET
Rows    : 229474
Columns : 22


,Diabetes_binary,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,HeartDiseaseorAttack,PhysActivity,Fruits,...,AnyHealthcare,NoDocbcCost,GenHlth,MentHlth,PhysHlth,DiffWalk,Sex,Age,Education,Income
0,0.0,1.0,1.0,1.0,40.0,1.0,0.0,0.0,0.0,0.0,...,1.0,0.0,5.0,18.0,15.0,1.0,0.0,9.0,4.0,3.0
1,0.0,0.0,0.0,0.0,25.0,1.0,0.0,0.0,1.0,0.0,...,0.0,1.0,3.0,0.0,0.0,0.0,0.0,7.0,6.0,1.0
2,0.0,1.0,1.0,1.0,28.0,0.0,0.0,0.0,0.0,1.0,...,1.0,1.0,5.0,30.0,30.0,1.0,0.0,9.0,4.0,8.0
3,0.0,1.0,0.0,1.0,27.0,0.0,0.0,0.0,1.0,1.0,...,1.0,0.0,2.0,0.0,0.0,0.0,0.0,11.0,3.0,6.0
4,0.0,1.0,1.0,1.0,24.0,0.0,0.0,0.0,1.0,1.0,...,1.0,0.0,2.0,3.0,0.0,0.0,0.0,11.0,5.0,4.0


In [5]:
# =============================================================================
# STEP 2: FEATURE AND TARGET SEPARATION
# =============================================================================


target_column = "Diabetes_binary"


X = df.drop(
    target_column,
    axis=1
)


y = df[target_column]



print("=" * 70)
print("FEATURE MATRIX (X)")
print("=" * 70)

print(f"Rows    : {X.shape[0]}")
print(f"Columns : {X.shape[1]}")


print("\n" + "=" * 70)
print("TARGET VARIABLE (y)")
print("=" * 70)

print(f"Rows : {y.shape[0]}")

display(y.head())

FEATURE MATRIX (X)
Rows    : 229474
Columns : 21

TARGET VARIABLE (y)
Rows : 229474


0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: Diabetes_binary, dtype: float64

In [6]:
# =============================================================================
# STEP 3: VERIFY TARGET DISTRIBUTION
# =============================================================================


print("=" * 70)
print("TARGET DISTRIBUTION")
print("=" * 70)


display(
    y.value_counts()
)


print("\nClass Percentage")


display(
    (y.value_counts(normalize=True)*100)
    .round(2)
)

TARGET DISTRIBUTION


Diabetes_binary
0.0    194377
1.0     35097
Name: count, dtype: int64


Class Percentage


Diabetes_binary
0.0    84.71
1.0    15.29
Name: proportion, dtype: float64

In [7]:
# =============================================================================
# STEP 4: FEATURE TYPE CHECK
# =============================================================================


print("=" * 70)
print("FEATURE DATA TYPES")
print("=" * 70)


display(
    X.dtypes.to_frame(
        name="Data Type"
    )
)

FEATURE DATA TYPES


,Data Type
HighBP,float64
HighChol,float64
CholCheck,float64
BMI,float64
Smoker,float64
Stroke,float64
HeartDiseaseorAttack,float64
PhysActivity,float64
Fruits,float64
Veggies,float64


In [8]:
# =============================================================================
# STEP 3: TRAIN-TEST SPLIT
# =============================================================================


from sklearn.model_selection import train_test_split


print("=" * 70)
print("TRAIN-TEST SPLIT")
print("=" * 70)


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


print("Training Data")
print("-" * 70)

print(f"X_train shape : {X_train.shape}")
print(f"y_train shape : {y_train.shape}")


print("\nTesting Data")
print("-" * 70)

print(f"X_test shape  : {X_test.shape}")
print(f"y_test shape  : {y_test.shape}")

TRAIN-TEST SPLIT
Training Data
----------------------------------------------------------------------
X_train shape : (183579, 21)
y_train shape : (183579,)

Testing Data
----------------------------------------------------------------------
X_test shape  : (45895, 21)
y_test shape  : (45895,)


In [9]:
# =============================================================================
# CLASS DISTRIBUTION AFTER SPLIT
# =============================================================================


print("=" * 70)
print("TRAINING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_train.value_counts(normalize=True)*100)
    .round(2)
)



print("\n" + "=" * 70)
print("TESTING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_test.value_counts(normalize=True)*100)
    .round(2)
)

TRAINING TARGET DISTRIBUTION


Diabetes_binary
0.0    84.71
1.0    15.29
Name: proportion, dtype: float64


TESTING TARGET DISTRIBUTION


Diabetes_binary
0.0    84.71
1.0    15.29
Name: proportion, dtype: float64

In [10]:
# =============================================================================
# STEP 4: FEATURE SCALING
# =============================================================================


from sklearn.preprocessing import StandardScaler


print("=" * 70)
print("FEATURE SCALING USING STANDARD SCALER")
print("=" * 70)


# Initialize scaler
scaler = StandardScaler()


# Fit only on training data
X_train_scaled = scaler.fit_transform(X_train)


# Transform test data using same scaler
X_test_scaled = scaler.transform(X_test)



print("Scaling completed successfully")


print("\n" + "=" * 70)
print("SHAPE AFTER SCALING")
print("=" * 70)


print(f"X_train_scaled shape : {X_train_scaled.shape}")
print(f"X_test_scaled shape  : {X_test_scaled.shape}")

FEATURE SCALING USING STANDARD SCALER
Scaling completed successfully

SHAPE AFTER SCALING
X_train_scaled shape : (183579, 21)
X_test_scaled shape  : (45895, 21)


In [11]:
# =============================================================================
# CONVERT SCALED DATA BACK TO DATAFRAME
# =============================================================================


X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns
)


X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns
)



print("=" * 70)
print("SCALED TRAIN DATA PREVIEW")
print("=" * 70)


display(X_train_scaled.head())

SCALED TRAIN DATA PREVIEW


,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,HeartDiseaseorAttack,PhysActivity,Fruits,Veggies,...,AnyHealthcare,NoDocbcCost,GenHlth,MentHlth,PhysHlth,DiffWalk,Sex,Age,Education,Income
0,1.096408,-0.889095,0.205176,-0.395435,1.072035,-0.215734,-0.338755,0.601961,-1.256327,0.509271,...,-4.187359,-0.3204,-0.563578,-0.455262,-0.517047,-0.477229,1.13290,-0.350440,0.020709,0.529942
1,1.096408,1.124739,0.205176,0.786628,-0.932805,-0.215734,-0.338755,0.601961,-1.256327,0.509271,...,0.238814,-0.3204,0.375330,-0.455262,-0.517047,-0.477229,-0.88269,0.620283,-0.986375,-1.381726
2,-0.912069,1.124739,0.205176,-0.690951,-0.932805,-0.215734,-0.338755,0.601961,0.795971,0.509271,...,0.238814,-0.3204,-0.563578,-0.455262,-0.517047,-0.477229,-0.88269,0.620283,1.027793,1.007859
3,-0.912069,1.124739,0.205176,-0.986467,-0.932805,-0.215734,-0.338755,0.601961,0.795971,0.509271,...,0.238814,-0.3204,-0.563578,-0.455262,-0.517047,-0.477229,-0.88269,1.591006,0.020709,-0.425892
4,1.096408,-0.889095,0.205176,0.195596,-0.932805,-0.215734,-0.338755,0.601961,0.795971,0.509271,...,0.238814,-0.3204,1.314239,-0.455262,0.698722,2.095428,-0.88269,-0.350440,-0.986375,-0.425892


In [12]:
# =============================================================================
# SCALING VERIFICATION
# =============================================================================


print("=" * 70)
print("MEAN AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.mean().round(2)
)



print("\n" + "=" * 70)
print("STANDARD DEVIATION AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.std().round(2)
)

MEAN AFTER SCALING


HighBP                  0.0
HighChol                0.0
CholCheck              -0.0
BMI                    -0.0
Smoker                 -0.0
Stroke                  0.0
HeartDiseaseorAttack    0.0
PhysActivity           -0.0
Fruits                 -0.0
Veggies                -0.0
HvyAlcoholConsump      -0.0
AnyHealthcare           0.0
NoDocbcCost             0.0
GenHlth                 0.0
MentHlth                0.0
PhysHlth                0.0
DiffWalk               -0.0
Sex                    -0.0
Age                     0.0
Education              -0.0
Income                 -0.0
dtype: float64


STANDARD DEVIATION AFTER SCALING


HighBP                  1.0
HighChol                1.0
CholCheck               1.0
BMI                     1.0
Smoker                  1.0
Stroke                  1.0
HeartDiseaseorAttack    1.0
PhysActivity            1.0
Fruits                  1.0
Veggies                 1.0
HvyAlcoholConsump       1.0
AnyHealthcare           1.0
NoDocbcCost             1.0
GenHlth                 1.0
MentHlth                1.0
PhysHlth                1.0
DiffWalk                1.0
Sex                     1.0
Age                     1.0
Education               1.0
Income                  1.0
dtype: float64

# ANNOVA

## 1. Baseline

In [13]:
# =============================================================================
# ANOVA BASELINE FEATURE SELECTION (TOP-K = 5,10,15,20)
# =============================================================================

from sklearn.feature_selection import SelectKBest, f_classif
import pandas as pd


# Fit ANOVA once with all features
anova_selector = SelectKBest(
    score_func=f_classif,

    k="all"
)

anova_selector.fit(
    X_train_scaled,
    y_train
)


# Create score dataframe
anova_scores = pd.DataFrame({

    "Feature": X_train_scaled.columns,

    "ANOVA Score": anova_selector.scores_

})


# Sort by ANOVA importance
anova_scores = anova_scores.sort_values(
    by="ANOVA Score",
    ascending=False
).reset_index(drop=True)



# Top-k values
k_values = [5, 10, 15, 20]

anova_results = {}


for k in k_values:

    top_features = (
        anova_scores
        .head(k)["Feature"]
        .tolist()
    )


    # Boolean mask for selecting columns
    mask = X_train_scaled.columns.isin(
        top_features
    )


    X_train_selected = X_train_scaled.loc[
        :, mask
    ]

    X_test_selected = X_test_scaled.loc[
        :, mask
    ]


    anova_results[k] = {

        "features": top_features,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "scores": anova_scores.head(k)

    }


    print("="*70)
    print(f"TOP {k} ANOVA FEATURES")
    print("="*70)

    print(top_features)

    print("\nScores:")
    print(anova_scores.head(k))



TOP 5 ANOVA FEATURES
['GenHlth', 'HighBP', 'BMI', 'DiffWalk', 'HighChol']

Scores:
    Feature   ANOVA Score
0   GenHlth  15028.459921
1    HighBP  12696.639767
2       BMI   8028.929585
3  DiffWalk   7990.290227
4  HighChol   7172.991333
TOP 10 ANOVA FEATURES
['GenHlth', 'HighBP', 'BMI', 'DiffWalk', 'HighChol', 'Age', 'HeartDiseaseorAttack', 'PhysHlth', 'Income', 'Education']

Scores:
                Feature   ANOVA Score
0               GenHlth  15028.459921
1                HighBP  12696.639767
2                   BMI   8028.929585
3              DiffWalk   7990.290227
4              HighChol   7172.991333
5                   Age   5995.341829
6  HeartDiseaseorAttack   5368.293621
7              PhysHlth   4532.827198
8                Income   3621.117911
9             Education   1906.303916
TOP 15 ANOVA FEATURES
['GenHlth', 'HighBP', 'BMI', 'DiffWalk', 'HighChol', 'Age', 'HeartDiseaseorAttack', 'PhysHlth', 'Income', 'Education', 'PhysActivity', 'Stroke', 'CholCheck', 'HvyAlcoholCo

In [14]:
# =============================================================================
# BASELINE MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {
    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=42
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        random_state=42,

        eval_metric="logloss",
        scale_pos_weight=(y_train == 0).sum() / (y_train == 1).sum()
    )
}

In [15]:
# =============================================================================
# MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

results = []

for k in k_values:

    Xtr = anova_results[k]["X_train"]
    Xts = anova_results[k]["X_test"]

    features = ", ".join(anova_results[k]["features"])

    for model_name, model in models.items():

        model.fit(Xtr, y_train)

        y_pred = model.predict(Xts)

        y_prob = model.predict_proba(Xts)[:,1]

        results.append({

            "Method": "ANOVA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(y_test, y_pred),

            "Precision": precision_score(y_test, y_pred),

            "Recall": recall_score(y_test, y_pred),

            "F1 Score": f1_score(y_test, y_pred),

            "ROC-AUC": roc_auc_score(y_test, y_prob)

        })

In [16]:
results_df = pd.DataFrame(results)

results_df

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,ANOVA,5,Logistic Regression,"GenHlth, HighBP, BMI, DiffWalk, HighChol",0.708552,0.308374,0.728736,0.433364,0.791913
1,ANOVA,5,Random Forest,"GenHlth, HighBP, BMI, DiffWalk, HighChol",0.703366,0.305606,0.738567,0.432324,0.786946
2,ANOVA,5,XGBoost,"GenHlth, HighBP, BMI, DiffWalk, HighChol",0.691077,0.299927,0.764496,0.430831,0.794595
3,ANOVA,10,Logistic Regression,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.713934,0.317044,0.754238,0.446431,0.805833
4,ANOVA,10,Random Forest,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.808912,0.337419,0.258869,0.292970,0.732218
5,ANOVA,10,XGBoost,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.710012,0.315326,0.765066,0.446588,0.807438
6,ANOVA,15,Logistic Regression,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.714653,0.317890,0.755663,0.447519,0.809178
7,ANOVA,15,Random Forest,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.824818,0.368867,0.204588,0.263196,0.754941
8,ANOVA,15,XGBoost,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.710470,0.316835,0.772475,0.449362,0.810981
9,ANOVA,20,Logistic Regression,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.714304,0.318260,0.760080,0.448659,0.810583


In [17]:
results_df.to_csv(
    "../../../results/diabetes/baseline/anova_baseline_results.csv",
    index=False
)

## 2. DODA

In [18]:
import sys
import subprocess

subprocess.check_call([
    sys.executable,
    "-m",
    "pip",
    "install",
    "--upgrade",
    "--force-reinstall",
    "--no-cache-dir",
    "git+https://github.com/anandha-3679/DODA.git"
])

0

In [19]:
import sys
import doda
from doda import DODASelector
import inspect

print("Python:", sys.executable)
print("DODA:", doda.__file__)
print("Selector:", inspect.getfile(DODASelector))

Python: c:\Users\Anandha Lakshmi\Downloads\Anaconda\python.exe
DODA: c:\Users\Anandha Lakshmi\Downloads\Anaconda\Lib\site-packages\doda\__init__.py
Selector: c:\Users\Anandha Lakshmi\Downloads\Anaconda\Lib\site-packages\doda\selector.py


In [20]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

from doda import DODASelector

from doda.adapters import SklearnAdapter

from doda.knowledge.providers import JSONProvider

from doda.fusion import HadamardFusion

In [21]:
# =============================================================================
# ANOVA + DODA FEATURE SELECTION
# =============================================================================

import pandas as pd
from sklearn.feature_selection import SelectKBest, f_classif

# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 10, 15, 20]


# -------------------------------------------------------------------------
# Clinical knowledge provider
# -------------------------------------------------------------------------

provider = JSONProvider(
    "clinical_weights.json"
)


# -------------------------------------------------------------------------
# Fusion method
# -------------------------------------------------------------------------

fusion = HadamardFusion()


# -------------------------------------------------------------------------
# Store DODA results
# -------------------------------------------------------------------------

doda_results = {}


# =============================================================================
# RUN DODA FOR EACH TOP-K
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"ANOVA + DODA — TOP {k} FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # ANOVA operator
    # ---------------------------------------------------------------------

    anova = SklearnAdapter(
        SelectKBest(
            score_func=f_classif,
            k=k
        )
    )


    # ---------------------------------------------------------------------
    # DODA selector
    # ---------------------------------------------------------------------

    selector = DODASelector(
        operators=[anova],
        provider=provider,
        fusion=fusion,
        top_k=k
    )


    # ---------------------------------------------------------------------
    # Fit DODA on TRAINING data
    # ---------------------------------------------------------------------

    X_train_selected = selector.fit_transform(
        X_train_scaled,
        y_train
    )


    # ---------------------------------------------------------------------
    # Transform TEST data
    # ---------------------------------------------------------------------

    X_test_selected = selector.transform(
        X_test_scaled
    )


    # ---------------------------------------------------------------------
    # Selected features
    # ---------------------------------------------------------------------

    selected_features = selector.get_selected_features()


    print("\nSelected Features:")
    print(selected_features)


    # ---------------------------------------------------------------------
    # Store everything needed for later evaluation
    # ---------------------------------------------------------------------

    doda_results[k] = {

        "selector": selector,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "features": selected_features,

        "raw_math_scores": selector.raw_math_scores_,

        "math_scores": selector.math_scores_,

        "clinical_weights": selector.clinical_weights_,

        "final_scores": selector.final_scores_
    }


print("\n" + "=" * 70)
print("DODA FEATURE SELECTION COMPLETED")
print("=" * 70)


ANOVA + DODA — TOP 5 FEATURES
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x000001F1BF1FF380>
DODA Pipeline Started
Running: SelectKBest
DODA feature order:
['HighBP', 'HighChol', 'CholCheck', 'BMI', 'Smoker', 'Stroke', 'HeartDiseaseorAttack', 'PhysActivity', 'Fruits', 'Veggies']
Number of features: 21
Number of scores: 21
Operator Scores:
{'SelectKBest_1': {'HighBP': np.float64(12696.639766679335), 'HighChol': np.float64(7172.991333444755), 'CholCheck': np.float64(963.9434241073692), 'BMI': np.float64(8028.929584555686), 'Smoker': np.float64(386.63520122466446), 'Stroke': np.float64(1781.4322362374398), 'HeartDiseaseorAttack': np.float64(5368.293621472076), 'PhysActivity': np.float64(1875.2111020627494), 'Fruits': np.float64(101.57045974595091), 'Veggies': np.float64(302.33873565463773), 'HvyAlcoholConsump': np.float64(804.7090266436618), 'AnyHealthcare': np.float64(133.8411247029314), 'NoDocbcCost': np.float64(48.478989334117266),

In [26]:
# CELL 2
# =============================================================================
# INSPECT MATHEMATICAL SCORES
# =============================================================================

k = 10

selector = doda_results[k]["selector"]

# -----------------------------------------------------------------------------
# Raw mathematical scores
# -----------------------------------------------------------------------------

raw_math_scores = pd.DataFrame(
    selector.raw_math_scores_.items(),
    columns=[
        "Feature",
        "Raw Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Normalized mathematical scores
# -----------------------------------------------------------------------------

normalized_math_scores = pd.DataFrame(
    selector.math_scores_.items(),
    columns=[
        "Feature",
        "Normalized Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Combine both
# -----------------------------------------------------------------------------

math_scores = raw_math_scores.merge(
    normalized_math_scores,
    on="Feature"
)

math_scores = math_scores.sort_values(
    by="Normalized Math Score",
    ascending=False
)

print("=" * 70)
print("MATHEMATICAL SCORES")
print("=" * 70)

display(math_scores)

MATHEMATICAL SCORES


,Feature,Raw Math Score,Normalized Math Score
13,GenHlth,15028.459921,1.000000
0,HighBP,12696.639767,0.844840
3,BMI,8028.929585,0.534248
16,DiffWalk,7990.290227,0.531677
1,HighChol,7172.991333,0.477294
18,Age,5995.341829,0.398933
6,HeartDiseaseorAttack,5368.293621,0.357208
15,PhysHlth,4532.827198,0.301616
20,Income,3621.117911,0.240951
19,Education,1906.303916,0.126846


In [27]:
# CELL 3
# =============================================================================
# INSPECT CLINICAL WEIGHTS
# =============================================================================

clinical_weights = pd.DataFrame(
    selector.clinical_weights_.items(),
    columns=[
        "Feature",
        "Clinical Weight"
    ]
)

clinical_weights = clinical_weights.sort_values(
    by="Clinical Weight",
    ascending=False
)

print("=" * 70)
print("CLINICAL WEIGHTS")
print("=" * 70)

display(clinical_weights)

CLINICAL WEIGHTS


,Feature,Clinical Weight
0,HighBP,1.0
3,BMI,1.0
13,GenHlth,1.0
6,HeartDiseaseorAttack,1.0
5,Stroke,1.0
18,Age,0.9
16,DiffWalk,0.9
1,HighChol,0.9
15,PhysHlth,0.8
4,Smoker,0.6


In [28]:
# CELL 4
# =============================================================================
# INSPECT FINAL DODA SCORES
# =============================================================================

final_scores = pd.DataFrame(
    selector.final_scores_.items(),
    columns=[
        "Feature",
        "Final DODA Score"
    ]
)

final_scores = final_scores.sort_values(
    by="Final DODA Score",
    ascending=False
)

print("=" * 70)
print("FINAL DODA SCORES")
print("=" * 70)

display(final_scores)

FINAL DODA SCORES


,Feature,Final DODA Score
13,GenHlth,1.000000
0,HighBP,0.844840
3,BMI,0.534248
16,DiffWalk,0.478510
1,HighChol,0.429564
18,Age,0.359039
6,HeartDiseaseorAttack,0.357208
15,PhysHlth,0.241293
5,Stroke,0.118537
20,Income,0.072285


In [29]:
selector.get_selected_features()

['GenHlth',
 'HighBP',
 'BMI',
 'DiffWalk',
 'HighChol',
 'Age',
 'HeartDiseaseorAttack',
 'PhysHlth',
 'Stroke',
 'Income']

In [30]:
comparison = selector.compare_scores()

comparison


,Feature,Math Rank,Normalized Math Score,Clinical Weight,DODA Rank,Final Score,Rank Change
0,GenHlth,1,1.0000,1.0,1,1.0000,0
1,HighBP,2,0.8448,1.0,2,0.8448,0
2,BMI,3,0.5342,1.0,3,0.5342,0
3,DiffWalk,4,0.5317,0.9,4,0.4785,0
4,HighChol,5,0.4773,0.9,5,0.4296,0
5,Age,6,0.3989,0.9,6,0.3590,0
6,HeartDiseaseorAttack,7,0.3572,1.0,7,0.3572,0
7,PhysHlth,8,0.3016,0.8,8,0.2413,0
8,Stroke,12,0.1185,1.0,9,0.1185,3
9,Income,9,0.2410,0.3,10,0.0723,-1


In [31]:
# =============================================================================
# DODA MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


doda_model_results = []


for k in k_values:

    Xtr = doda_results[k]["X_train"]

    Xts = doda_results[k]["X_test"]

    features = ", ".join(
        doda_results[k]["features"]
    )


    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        doda_model_results.append({

            "Method": "ANOVA + DODA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


doda_results_df = pd.DataFrame(
    doda_model_results
)


display(doda_results_df)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,ANOVA + DODA,5,Logistic Regression,"GenHlth, HighBP, BMI, DiffWalk, HighChol",0.708552,0.308374,0.728736,0.433364,0.791913
1,ANOVA + DODA,5,Random Forest,"GenHlth, HighBP, BMI, DiffWalk, HighChol",0.703366,0.305629,0.738709,0.432372,0.786912
2,ANOVA + DODA,5,XGBoost,"GenHlth, HighBP, BMI, DiffWalk, HighChol",0.691077,0.299927,0.764496,0.430831,0.794595
3,ANOVA + DODA,10,Logistic Regression,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.714283,0.317720,0.756660,0.447525,0.805912
4,ANOVA + DODA,10,Random Forest,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.793703,0.320533,0.311583,0.315995,0.718421
5,ANOVA + DODA,10,XGBoost,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.708857,0.314543,0.766348,0.446020,0.807959
6,ANOVA + DODA,15,Logistic Regression,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.712975,0.316255,0.754523,0.445697,0.807390
7,ANOVA + DODA,15,Random Forest,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.829546,0.380995,0.183359,0.247571,0.758248
8,ANOVA + DODA,15,XGBoost,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.710753,0.316303,0.767346,0.447956,0.808930
9,ANOVA + DODA,20,Logistic Regression,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.714304,0.318260,0.760080,0.448659,0.810583


In [32]:
# =============================================================================
# COMBINED COMPARISON
# =============================================================================

comparison_df = pd.concat(
    [
        results_df,
        doda_results_df
    ],
    ignore_index=True
)


display(
    comparison_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,ANOVA,5,Logistic Regression,"GenHlth, HighBP, BMI, DiffWalk, HighChol",0.708552,0.308374,0.728736,0.433364,0.791913
1,ANOVA,5,Random Forest,"GenHlth, HighBP, BMI, DiffWalk, HighChol",0.703366,0.305606,0.738567,0.432324,0.786946
2,ANOVA,5,XGBoost,"GenHlth, HighBP, BMI, DiffWalk, HighChol",0.691077,0.299927,0.764496,0.430831,0.794595
3,ANOVA,10,Logistic Regression,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.713934,0.317044,0.754238,0.446431,0.805833
4,ANOVA,10,Random Forest,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.808912,0.337419,0.258869,0.292970,0.732218
5,ANOVA,10,XGBoost,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.710012,0.315326,0.765066,0.446588,0.807438
6,ANOVA,15,Logistic Regression,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.714653,0.317890,0.755663,0.447519,0.809178
7,ANOVA,15,Random Forest,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.824818,0.368867,0.204588,0.263196,0.754941
8,ANOVA,15,XGBoost,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.710470,0.316835,0.772475,0.449362,0.810981
9,ANOVA,20,Logistic Regression,"GenHlth, HighBP, BMI, DiffWalk, HighChol, Age,...",0.714304,0.318260,0.760080,0.448659,0.810583


In [33]:
for k in k_values:

    selector = doda_results[k]["selector"]

    raw_rank = [
        x[0]
        for x in sorted(
            selector.raw_math_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    doda_rank = [
        x[0]
        for x in sorted(
            selector.final_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    print("\n" + "=" * 70)
    print(f"TOP {k}")
    print("=" * 70)

    print("Raw Math Ranking:")
    print(raw_rank)

    print("\nDODA Ranking:")
    print(doda_rank)

    print("\nTop-K Same:",
          raw_rank[:k] == doda_rank[:k])


TOP 5
Raw Math Ranking:
['GenHlth', 'HighBP', 'BMI', 'DiffWalk', 'HighChol', 'Age', 'HeartDiseaseorAttack', 'PhysHlth', 'Income', 'Education', 'PhysActivity', 'Stroke', 'CholCheck', 'HvyAlcoholConsump', 'MentHlth', 'Smoker', 'Veggies', 'Sex', 'AnyHealthcare', 'Fruits', 'NoDocbcCost']

DODA Ranking:
['GenHlth', 'HighBP', 'BMI', 'DiffWalk', 'HighChol', 'Age', 'HeartDiseaseorAttack', 'PhysHlth', 'Stroke', 'Income', 'PhysActivity', 'Education', 'CholCheck', 'MentHlth', 'Smoker', 'HvyAlcoholConsump', 'Veggies', 'Sex', 'AnyHealthcare', 'Fruits', 'NoDocbcCost']

Top-K Same: True

TOP 10
Raw Math Ranking:
['GenHlth', 'HighBP', 'BMI', 'DiffWalk', 'HighChol', 'Age', 'HeartDiseaseorAttack', 'PhysHlth', 'Income', 'Education', 'PhysActivity', 'Stroke', 'CholCheck', 'HvyAlcoholConsump', 'MentHlth', 'Smoker', 'Veggies', 'Sex', 'AnyHealthcare', 'Fruits', 'NoDocbcCost']

DODA Ranking:
['GenHlth', 'HighBP', 'BMI', 'DiffWalk', 'HighChol', 'Age', 'HeartDiseaseorAttack', 'PhysHlth', 'Stroke', 'Income', 'P

## ANOVA + DODA: Model Evaluation Interpretation

The initial evaluation was performed using **ANOVA as the mathematical feature selection method**, followed by DODA-based clinical feature adjustment. This represents only the first stage of the overall experimental framework; the same evaluation will subsequently be performed with three additional feature selection methods.

### 1. Overall Predictive Performance

The results show that **ANOVA + DODA does not produce a consistent improvement in predictive performance across all models and Top-K values**.

For Logistic Regression and XGBoost, the performance of ANOVA and ANOVA + DODA is generally very similar. For example, at Top-20, both approaches produce an F1-score of approximately **0.449 for Logistic Regression** and **0.452 for XGBoost**. This indicates that the DODA-based clinical adjustment does not substantially alter predictive performance when the resulting selected feature set remains similar.

Random Forest shows a more noticeable variation. At Top-10, ANOVA + DODA improves the F1-score from **0.2930 to 0.3160**, while at Top-15 the F1-score decreases from **0.2632 to 0.2476**. Therefore, the effect of DODA on predictive performance is **model- and Top-K-dependent rather than uniformly positive**.

### 2. Accuracy Should Not Be Considered in Isolation

The Random Forest results demonstrate an important issue with relying only on accuracy. For example, at Top-20:

- ANOVA: Accuracy = **0.8424**
- ANOVA + DODA: Accuracy = **0.8427**

Although accuracy is high, recall is only approximately **0.15**, indicating that the model identifies relatively few positive cases.

In contrast, Logistic Regression and XGBoost generally achieve lower accuracy but substantially higher recall and F1-score. Therefore, because this is a **clinical prediction problem with class imbalance**, F1-score, recall, precision, and ROC-AUC provide more informative measures of model behavior than accuracy alone.

### 3. Effect of DODA on Feature Selection

The current results show that the selected feature lists for ANOVA and ANOVA + DODA are largely identical or highly similar across the evaluated Top-K values. Consequently, it is expected that many of the downstream model results are also identical or very close.

This does **not necessarily indicate that DODA provides no benefit**. DODA modifies the mathematical feature scores by incorporating clinical knowledge before the final ranking. If the clinical adjustment changes feature scores but does not change the Top-K cutoff, the downstream model receives the same features and therefore its predictive performance may remain unchanged.

Thus, predictive performance alone is insufficient to evaluate the contribution of DODA. The **change in feature ranking and clinical relevance of selected features must also be evaluated**.

### 4. Importance of the New Normalization Step

The updated DODA framework introduces normalization between the mathematical feature-selection scores and the clinical knowledge layer:

**Raw Mathematical Scores → Normalization → Clinical Weights → Fusion → Final DODA Ranking**

This ensures that mathematical feature-selection scores are placed on a comparable scale before they are combined with clinical weights. The raw mathematical scores are retained separately to provide an auditable comparison between the original feature-selection scores and the normalized scores used for fusion.

Therefore, the DODA experiments should be **rerun using the updated framework** before the final results are reported.

### 5. Current Conclusion

Based on the ANOVA experiment alone, **DODA maintains broadly comparable predictive performance to conventional ANOVA while introducing an additional clinical knowledge-based ranking mechanism**. However, the current results do not demonstrate a consistent improvement in accuracy, F1-score, recall, or ROC-AUC across all models and Top-K values.

The more important question for the overall study is whether DODA can **improve the clinical relevance and stability of feature selection without sacrificing predictive performance**.

Since ANOVA represents only one mathematical feature-selection method, these findings should be considered **preliminary**. The same analysis will be repeated using the remaining three feature-selection methods. Comparing the behavior of DODA across different mathematical selectors will determine whether the observed effects are specific to ANOVA or represent a consistent property of the DODA framework.

### Key Findings from the ANOVA Experiment

| Finding | Interpretation |
|---|---|
| DODA does not consistently improve model metrics | Its primary contribution may not be raw predictive improvement |
| Logistic Regression and XGBoost show very similar performance | DODA generally preserves predictive behavior for these models |
| Random Forest shows larger changes | DODA's effect can depend on the downstream model and Top-K |
| Several ANOVA and DODA feature sets are similar | Explains the small differences in model performance |
| Accuracy is high for Random Forest but recall is low | Accuracy alone is misleading under class imbalance |
| Normalization is now applied before clinical fusion | Provides a standardized basis for combining mathematical and clinical scores |
| Raw scores are retained | Enables transparent comparison of pre- and post-normalization rankings |
| More FS methods remain to be evaluated | Current findings are preliminary and should not be generalized yet |

**Overall:** The ANOVA experiment suggests that DODA can incorporate clinical knowledge into feature ranking while generally preserving predictive performance, but it does not yet provide evidence of a universal predictive-performance advantage. The remaining feature-selection methods are necessary to determine the robustness and generalizability of this observation.